django/utils/numberformat.py

```diff
@@ -18,38 +18,58 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
     * grouping: Number of digits in every group limited by thousand separator
     * thousand_sep: Thousand separator symbol (for example ",")
     """
-    use_grouping = settings.USE_L10N and settings.USE_THOUSAND_SEPARATOR
-    use_grouping = use_grouping or force_grouping
-    use_grouping = use_grouping and grouping > 0
-    # Make the common case fast
-    if isinstance(number, int) and not use_grouping and not decimal_pos:
-        return mark_safe(six.text_type(number))
-    # sign
-    sign = ''
+    use_grouping = grouping > 0 and (force_grouping or
+                                     (settings.USE_THOUSAND_SEPARATOR and settings.USE_L10N))
+    if not use_grouping or grouping == 3:
+        try:
+            # Make the common case fast
+            #
+            # Works on numbers when the formatting can be handled by the built in format
+            # function (e.g. grouping is disabled or 3).
+            #
+            # Will also work on strings when grouping is disabled and decimal_pos is None.
+            if decimal_pos is not None:
+                # Add one toe decimal_pos to prevent visible rounding. For example:
+                # If `number` is 10.888 and `decimal_pos` is 2 return 10.88 not 10.89.
+                format_string = '.%sf' % (decimal_pos + 1)
+            elif isinstance(number, Decimal):
+                format_string = 'f'
+            else:
+                format_string = ''
+            if use_grouping and (number >= 1000 or number <= -1000):
+                format_string = ',' + format_string  # PEP 0378
```

Comment:
I think more comments are needed throughout this. I don't have all PEPs memorized, so just stating the number isn't so helpful as I still have to look it up.